ports/mimxrt10xx/common-hal/audiobusio/__init__.c

```diff
@@ -0,0 +1,422 @@
+                        &self->sample_data, &sample_buffer_length);
+                self->sample_end = self->sample_data + sample_buffer_length;
+                if (get_buffer_result == GET_BUFFER_DONE) {
+                    if (self->loop) {
+                        audiosample_reset_buffer(self->sample, false, 0);
+                    } else {
+                        self->stopping = true;
+                    }
+                }
+                if (get_buffer_result == GET_BUFFER_ERROR || sample_buffer_length == 0) {
+                    self->stopping = true;
+                }
+            }
+            size_t input_bytecount = self->sample_end - self->sample_data;
+            size_t bytes_per_input_frame = self->channel_count * self->bytes_per_sample;
+            size_t framecount = MIN((size_t)(end - ptr), input_bytecount / bytes_per_input_frame);
+
+#define SAMPLE_TYPE(is_signed, channel_count, bytes_per_sample) ((is_signed) | ((channel_count) << 1) | ((bytes_per_sample) << 3))
+
+            switch (SAMPLE_TYPE(self->samples_signed, self->channel_count, self->bytes_per_sample)) {
+
+                case SAMPLE_TYPE(true, 2, 2):
+                    memcpy(ptr, self->sample_data, 4 * framecount);
+                    break;
+
+                case SAMPLE_TYPE(false, 2, 2):
+                    audiosample_convert_u16s_s16s((int16_t *)ptr, (uint16_t *)(void *)self->sample_data, framecount);
+                    break;
+
+                case SAMPLE_TYPE(true, 1, 2):
+                    audiosample_convert_s16m_s16s((int16_t *)ptr, (int16_t *)(void *)self->sample_data, framecount);
+                    break;
+
+                case SAMPLE_TYPE(false, 1, 2):
+                    audiosample_convert_u16m_s16s((int16_t *)ptr, (uint16_t *)(void *)self->sample_data, framecount);
+                    break;
+
+                case SAMPLE_TYPE(true, 2, 1):
+                    audiosample_convert_s8s_s16s((int16_t *)ptr, (int8_t *)(void *)self->sample_data, framecount);
+                    memcpy(ptr, self->sample_data, 4 * framecount);
+                    break;
+
+                case SAMPLE_TYPE(false, 2, 1):
+                    audiosample_convert_u8s_s16s((int16_t *)ptr, (uint8_t *)(void *)self->sample_data, framecount);
+                    break;
+
+                case SAMPLE_TYPE(true, 1, 1):
+                    audiosample_convert_s8m_s16s((int16_t *)ptr, (int8_t *)(void *)self->sample_data, framecount);
+                    break;
+
+                case SAMPLE_TYPE(false, 1, 1):
+                    audiosample_convert_u8m_s16s((int16_t *)ptr, (uint8_t *)(void *)self->sample_data, framecount);
+                    break;
+            }
+            self->sample_data += bytes_per_input_frame * framecount; // in bytes
+            ptr += framecount; // in frames
+        }
+        // Fill any remaining portion of the buffer with 'no sound'
+        memset(ptr, 0, (end - ptr) * sizeof(uint32_t));
```

Comment:
Don't we allow the user to set the quiescent value? Or is it ok that it is always signed?